contracts/zero-ex/contracts/src/features/UniswapV2ForkFeature.sol

```diff
@@ -0,0 +1,205 @@
+    )
+        internal
+        view
+        returns (uint[] memory amounts)
+    {
+        require(tokens.length >= 2, 'UniswapV2ForkFeature/INVALID_PATH');
+
+        amounts = new uint[](tokens.length);
+        amounts[0] = amountIn;
+        for (uint i; i < tokens.length - 1; i++) {
+            (uint reserveIn, uint reserveOut) = _getReserves(factory, tokens[i], tokens[i + 1]);
+            amounts[i + 1] = _getAmountOut(amounts[i], reserveIn, reserveOut);
+        }
+    }
+
+    function _getReserves(
+        address factory,
+        IERC20TokenV06 tokenA,
+        IERC20TokenV06 tokenB
+    )
+        internal
+        view
+        returns (uint reserveA, uint reserveB)
+    {
+        (IERC20TokenV06 token0,) = _sortTokens(tokenA, tokenB);
+        (uint reserve0, uint reserve1,) = IUniswapV2ForkPair(_pairFor(factory, tokenA, tokenB)).getReserves();
+        (reserveA, reserveB) = tokenA == token0 ? (reserve0, reserve1) : (reserve1, reserve0);
+    }
+
+    function _sortTokens(
+        IERC20TokenV06 tokenA,
+        IERC20TokenV06 tokenB
+    )
+        internal
+        pure
+        returns (IERC20TokenV06 token0, IERC20TokenV06 token1)
+    {
+        require(tokenA != tokenB, 'UniswapV2ForkFeature/IDENTICAL_ADDRESSES');
+        (token0, token1) = tokenA < tokenB ? (tokenA, tokenB) : (tokenB, tokenA);
+        require(token0 != IERC20TokenV06(0), 'UniswapV2ForkFeature/ZERO_ADDRESS');
+    }
+
+    function _getAmountOut(
+        uint amountIn,
+        uint reserveIn,
+        uint reserveOut
+    )
+        internal
+        pure
+        returns (uint amountOut)
+    {
+        require(amountIn > 0, 'UniswapV2ForkFeature/INSUFFICIENT_INPUT_AMOUNT');
+        require(reserveIn > 0 && reserveOut > 0, 'UniswapV2ForkFeature/INSUFFICIENT_LIQUIDITY');
+        uint amountInWithFee = amountIn.safeMul(997);
+        uint numerator = amountInWithFee.safeMul(reserveOut);
+        uint denominator = reserveIn.safeMul(1000).safeMul(amountInWithFee);
+        amountOut = numerator / denominator;
+    }
+
+    function _getFactory(ProtocolFork fork) private view returns (address) {
+        // TODO: complete the list
+        if (fork == ProtocolFork.SushiSwap) {
+            return SUSHISWAPV2_FACTORY;
+        } else {
+            revert('UniswapV2ForkFeature/UNSUPPORTED_FORK');
+        }
+    }
+
+    function _pairFor(
+        address factory,
+        IERC20TokenV06 tokenA,
+        IERC20TokenV06 tokenB
+    )
+        internal
+        pure
+        returns (address pair)
+    {
+        (IERC20TokenV06 token0, IERC20TokenV06 token1) = _sortTokens(tokenA, tokenB);
+        pair = address(uint(keccak256(abi.encodePacked(
+            hex'ff',
+            factory,
+            keccak256(abi.encodePacked(address(token0), address(token1))),
+            hex'96e8ac4277198ff8b6f785478aa9a39f403cb768dd02cbee326c3e7da348845f' // init code hash TODO(Cece): how can I find the init code for each fork
```

Comment:
Usually their router contract will have the init code hash somewhere in their code.